quantization/src/encoded_vectors_pq.rs

```diff
@@ -0,0 +1,315 @@
+
+            for (i, vector_data) in data.clone().into_iter().enumerate() {
+                let subvector_data = &vector_data[range.clone()];
+                let centroid_index =
+                    storage.get_vector_data(i, vector_division.len())[range_i] as usize;
+                centroids_counter[centroid_index] += 1;
+                let x = (((subvector_data[0] - min) / (max - min)) * imgx as f32)
+                    .clamp(0., imgx as f32 - 1.0) as u32;
+                let y = (((subvector_data[1] - min) / (max - min)) * imgy as f32)
+                    .clamp(0., imgy as f32 - 1.0) as u32;
+                *imgbuf.get_pixel_mut(x, y) = image::Rgb([
+                    colors_r[centroid_index],
+                    colors_g[centroid_index],
+                    colors_b[centroid_index],
+                ]);
+            }
+
+            for centroid in centroids {
+                let subvector_data = &centroid[range.clone()];
+                let x = (((subvector_data[0] - min) / (max - min)) * imgx as f32)
+                    .clamp(0., imgx as f32 - 2.0) as u32;
+                let y = (((subvector_data[1] - min) / (max - min)) * imgy as f32)
+                    .clamp(0., imgy as f32 - 2.0) as u32;
+                *imgbuf.get_pixel_mut(x, y) = image::Rgb([255u8, 0u8, 0u8]);
+                *imgbuf.get_pixel_mut(x + 1, y) = image::Rgb([255u8, 0u8, 0u8]);
+                *imgbuf.get_pixel_mut(x + 1, y + 1) = image::Rgb([255u8, 0u8, 0u8]);
+                *imgbuf.get_pixel_mut(x, y + 1) = image::Rgb([255u8, 0u8, 0u8]);
+            }
+
+            imgbuf.save(&format!("kmeans-{range_i}.png")).unwrap();
+        }
+    }
+}
+
+impl<TStorage: EncodedStorage> EncodedVectors<EncodedQueryPQ> for EncodedVectorsPQ<TStorage> {
+    fn save(&self, data_path: &Path, meta_path: &Path) -> std::io::Result<()> {
+        let metadata_bytes = serde_json::to_vec(&self.metadata)?;
+        meta_path.parent().map(std::fs::create_dir_all);
+        let mut buffer = File::create(meta_path)?;
+        buffer.write_all(&metadata_bytes)?;
+
+        data_path.parent().map(std::fs::create_dir_all);
+        self.encoded_vectors.save_to_file(data_path)?;
+        Ok(())
+    }
+
+    fn load(
+        data_path: &Path,
+        meta_path: &Path,
+        vector_parameters: &VectorParameters,
+    ) -> std::io::Result<Self> {
+        let mut contents = String::new();
+        let mut file = File::open(meta_path)?;
+        file.read_to_string(&mut contents)?;
+        let metadata: Metadata = serde_json::from_str(&contents)?;
+        let quantized_vector_size = metadata.vector_division.len();
+        let encoded_vectors =
+            TStorage::from_file(data_path, quantized_vector_size, vector_parameters.count)?;
+        let result = Self {
+            metadata,
+            encoded_vectors,
+        };
+        Ok(result)
+    }
+
+    fn encode_query(&self, query: &[f32]) -> EncodedQueryPQ {
+        let mut lut = Vec::new();
```

Comment:
At this point we should know how long exactly the lut will be. Might make sense to pre-allocate.

Additionally, we can use rotating buffers to avoid allocation completely

Reply:
Fixed prealloc. I'm against rotating buffer because raw scorer creation is doing once per search. Even for original raw scorer `Vec<f32>` is allocated each time